docs/source/features/optimizations/communication_overlap.rst

```diff
@@ -1,5 +1,5 @@
 Communication Overlap
```

Comment:
Technical edit of communication_overlap.rst by line numbers.

Line 3. Need to add summary text for this topic. Something like:

Communication overlap strategies are crucial for training large-scale models efficiently, as they help to utilize computational resources better and reduce training time. This section describes strategies for overlapping the data-parallel (DP), tensor-parallel (TP), and pipeline-parallel (PP) communications with computation operations.

Lines 7-8 suggested revision.

NeMo enables the overlap of data-parallel (DP) communications with computations during LLM training. It also includes a Distributed Optimizer that distributes optimizer states and high-precision master parameters across GPUs.

Line 9. fix capitalization.

The DP communication is chunked by the granularity of a transformer layer and overlaps each communication chunk with computation.

Line 10. fix punctuation.

This overlap method exposes only one DP communication chunk, ensuring efficient large-scale LLM training.

Lines 13 - 15. suggested revision.

DP gradient reduce-scatter and parameter all-gather overlaps are enabled when setting ``overlap_grad_sync=true`` and ``overlap_param_sync=true``, respectively.The precision of the gradient reduce-scatter is set by ``grad_sync_dtype``. Reducing  bf16 ensures improved performance at large scale training compared to the default precision of fp32.
When training in fp8 computing precision (with ``fp8=true``), setting ``fp8_params=true`` conducts the parameter all-gather in fp8 thereby reducing the all-gather overhead by half.

Line 20. revise.

Tensor parallelism, used with the sequence-parallel activation sharding (``sequence_parallel=true``), introduces activation (gradient) all-gather and reduce-scatter as shown in the figure below.

Line 22. revise. 

The TP communication without direct computation dependency is overlapped with the computation in bulk (the linear layer and TP communication pairs in the yellow boxes).

Line 25. revise.

The TP communication and computation are chunked and the chunks are overlapped in the pipeline.

Line 26. revise run-on sentence.

In the pipelined overlap, the activation (gradient) tensor all-gather is replaced with multiple steps of input P2P ring exchanges. Similarly, reduce-scatter is replaced with multiple steps of GEMM output P2P ring exchanges followed by a reduction of the received outputs.

Lines 29 - 32. Please check .. image:: ../nlp/nemo_megatron/images/tp_comm_overlap.png. It is not showing up in the HTML file.

Lines 34 -36. revise.

The pipelined TP communication overlap is implemented in Transformer Engine and enabled by setting ub_tp_comm_overlap=true. The specific overlap methods can be set in a config dictionary and passed as a YAML file. The individual bulk, pipelined all-gather, and reduce-scatter operations can be enabled or disabled by using tp_comm_bulk_wgrad, tp_comm_bulk_dgrad, tp_comm_overlap_ag, and tp_comm_overlap_rs, respectively.

Lines 41 - 43. Please check the original sentence: "This increasing PP communication overhead and it cancels off the reduced the pipeline bubbles with virtual pipelining." Here is my suggested revision.

Pipelining enables P2P activation (gradient) sends and receives between GPUs in a pipeline-parallel (PP) configuration. As the virtual pipeline parallel size increases, the frequency of PP communications also increases due to a reduction in the number of transformer layers processed per micro-batch. However, this increase in PP communication overhead counteracts the reduction in pipeline stalls that virtual pipelining is supposed to mitigate.

Line 52. revise.

The PP communication overlap is enabled by setting ``overlap_p2p_comm=true``. In addition, setting ``batch_p2p_comm=false`` uses separate kernels for the send and the receive operations, which further improves the communication efficiency and GPU resource utilization.

Line 59. suggested revision.

Context parallelism divides the activations (gradients) across all layers within the sequence domain. This division introduces all-gather and reduce-scatter operations for activations (gradients) during the forward and backward propagations of self-attention mechanisms.


Line 63. revise.

By default, the CP communication overlap is enabled when context parallelism is used (``context_parallel_size > 1``).